### PR TITLE
Partition tests on CI for faster execution

### PR DESCRIPTION
**Pull request by klochowicz**

### CODE REVIEW

**Comment by klochowicz** [comment]
unfortunately, `nextest` seems to run into some troubles with `otel-tests-macro` crate. I'm parking this until we extract some of these crates out of the workspace.

**Review by holzeis — APPROVED** [review]
Nice 👍

**Comment by klochowicz** [comment]
@da-kami have a look at this. after rust-cache settling in for the new jobs, the worst case for CI run on the new jobs was:

8m+8m = 16m (build ubuntu + run-tests-partitioned).

This is still heaps faster than the fastest runs on the pending PRs at the moment - some sample frontend dependency updates still go for 24m. 

This means it reduces the time by 1/3.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -98,10 +98,40 @@ jobs:
...
+          name: nextest-archive-${{ matrix.os }}
+          path: nextest-archive.tar.zst
+
+  run-tests-partitioned:
+    name: Run tests with partitioning
+    strategy:
+      matrix:
+        partition: [1, 2]
+        os: [ubuntu-latest, macos-12]
+    # XXX: Unfortunately github actions do not support matrix in 'needs' section
```
Comment (luckysori) [inline .github/workflows/ci.yml]: 🙃 I don't fully understand what to do with this comment. Can we just remove it or make it clearer?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -98,10 +98,40 @@ jobs:
...
-        run: cargo nextest run --profile ci
+      - name: Build and archive tests
+        run: cargo nextest archive --archive-file nextest-archive.tar.zst
+      - name: Upload archive to workflow
+        uses: actions/upload-artifact@v3
+        with:
+          name: nextest-archive-${{ matrix.os }}
+          path: nextest-archive.tar.zst
+
+  run-tests-partitioned:
```
Comment (luckysori) [inline .github/workflows/ci.yml]: ❓ Do we need a new name for this?

I would have opted for the shorter `build-test` and `test` (or `run-test`).

**Review by luckysori — APPROVED** [review]
Let's give it a go, nice idea!

Can you squash the commits into one once you've decided how many partitions we want?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (klochowicz) [reply .github/workflows/ci.yml]: changed to `build-tests` and `run-tests`

File: .github/workflows/ci.yml (hunk 1 above)
Comment (klochowicz) [reply .github/workflows/ci.yml]: removed the comment; it turned out not essential for writing a working GH action.

**Comment by klochowicz** [comment]
We seem to be hitting concurrency limits with macos-12 job, I've created #2943 to see whether this will alleviate the issue.

**Comment by klochowicz** [comment]
@da-kami @bonomat are you OK with me merging this?

**Review by luckysori — APPROVED** [review]

**Review by da-kami — APPROVED** [review]
Nice that we can achieve this with automated partitioning, i.e. we don't have to specify the partitions by hand. The fact that we have to specify the `4` partitions for bors is somewhat annoying, but I guess we can't work around that.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -98,10 +99,40 @@ jobs:
...
+      matrix:
+        partition: [1, 2, 3, 4]
+        os: [ubuntu-latest, macos-12]
+    needs: build-tests
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v3
+      # In this example, the Rust toolchain is not installed. cargo nextest's run phase does not
+      # require Cargo. You can choose to install Rust if your tests require it, of course.
+      # Instead, create ~/.cargo/bin as required by the install action.
```
Comment (bonomat) [inline .github/workflows/ci.yml:123]: This comment could have been rewritten slightly :)

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -98,10 +99,40 @@ jobs:
...
+      - run: mkdir -p ~/.cargo/bin
+      - name: Install nextest
+        uses: taiki-e/install-action@nextest
+      - name: Download archive
+        uses: actions/download-artifact@v3
+        with:
+          name: nextest-archive-${{ matrix.os }}
+      - name: Run tests
+        run: |
+          ~/.cargo/bin/cargo-nextest nextest run --archive-file nextest-archive.tar.zst \
```
Comment (bonomat) [inline .github/workflows/ci.yml:133]: Do you need to specify the whole path?

**Comment by bonomat** [comment]
> @da-kami @bonomat are you OK with me merging this?

Sorry for slow response. Looks good to me.

**Pull request merged**